File: src/sql/postgres/protocol/BackendKeyData.zig
const BackendKeyData = @This();

process_id: u32 = 0,
secret_key: u32 = 0,
pub const decode = DecoderWrap(BackendKeyData, decodeInternal).decode;

pub fn decodeInternal(this: *@This(), comptime Container: type, reader: NewReader(Container)) !void {
    if (!try reader.expectInt(u32, 12)) {
        return error.InvalidBackendKeyData;
    }

    this.* = .{
        .process_id = @bitCast(try reader.int4()),
        .secret_key = @bitCast(try reader.int4()),
    };
}

const DecoderWrap = @import("./DecoderWrap.zig").DecoderWrap;

const NewReader = @import("./NewReader.zig").NewReader;

File: src/sql/postgres/protocol/BackendKeyData.rs
use super::new_reader::NewReader;
use crate::postgres::AnyPostgresError;
use crate::postgres::types::int_types::Int4;

pub struct BackendKeyData {
    pub process_id: Int4,
    pub secret_key: Int4,
}

impl BackendKeyData {
    pub fn decode_internal<Container: super::new_reader::ReaderContext>(
        mut reader: NewReader<Container>,
    ) -> Result<Self, AnyPostgresError> {
        if !reader.expect_int::<u32>(12)? {
            return Err(AnyPostgresError::InvalidBackendKeyData);
        }

        Ok(Self {
            process_id: reader.int4()?,
            secret_key: reader.int4()?,
        })
    }
}
